pennylane/noise/noise_model.py

```diff
@@ -0,0 +1,156 @@
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains class and methods for noise models"""
+
+import inspect
+
+import pennylane as qml
+
+
+class NoiseModel:
+    """Builds a noise model based on a mapping of conditionals to callables that
+    defines noise operations using some optional metadata.
+
+    Args:
+        model_map (dict[~.BooleanFn -> Callable]): Data for the
+            noise model as a ``{conditional: noise_fn}`` dictionary. The signature of
+            ``noise_fn`` must be ``noise_fn(op: Operation, **kwargs) -> None``, where
+            ``op`` is the operation that the conditional evaluates and ``kwargs`` are
+            the specified metadata arguments.
+        **kwargs: Keyword arguments for specifying metadata related to noise model.
+
+    .. note::
+
+        For each key-value pair of ``model_map``:
+
+        - The ``conditional`` should be either a function decorated with :class:`~.BooleanFn`,
+          a callable object built via the constructor functions (:func:`~.pennylane.noise.op_eq`,
+          :func:`~.pennylane.noise.op_in`, :func:`~.pennylane.noise.wires_eq`, and
+          :func:`~.pennylane.noise.wires_in`), or their bit-wise combination.
+        - Defintion of ``noise_fn(op, **kwargs)`` should have the operations in same the order
+          in which they are to be queued for an operation ``op``, whenever the corresponding
+          ``conditional`` evaluates to ``True``.
+
+    **Example**
+
+    .. code-block:: python
+
+        # Set up the conditionals
+        c0 = qml.noise.op_eq(qml.PauliX) | qml.noise.op_eq(qml.PauliY)
+        c1 = qml.noise.op_eq(qml.Hadamard) & qml.noise.wires_in([0, 1])
+        c2 = qml.noise.op_eq(qml.RX)
+
+        @qml.BooleanFn
+        def c3(op, **kwargs):
+            return isinstance(op, qml.RY) and op.parameters[0] >= 0.5
+
+        # Set up the noise functions
+        n0 = qml.noise.partial_wires(qml.AmplitudeDamping, 0.4)
+
+        def n1(op, **kwargs):
+            ThermalRelaxationError(0.4, kwargs["t1"], 0.2, 0.6, op.wires)
+
+        def n2(op, **kwargs):
+            qml.RX(op.parameters[0] * 0.05, op.wires)
+
+        n3 = qml.noise.partial_wires(qml.PhaseDamping, 0.9)
+
+        # Set up noise model
+        noise_model = qml.NoiseModel({c0: n0, c1: n1, c2: n2}, t1=0.04)
+        noise_model += {c3: n3}
+
+    >>> noise_model
+    NoiseModel({
+        OpEq(PauliX) | OpEq(PauliY): AmplitudeDamping(0.4, wires),
```

Comment:
I like the ":" but when I run it I get "="

Reply:
Fixed!